In [1]:
from scipy.stats import chisquare 
import numpy as np

# Задача 1

Напиши функцию, которая удалит дубликаты в списке. Список не отсортирован. Необходимо сохранить порядок сортировки оригинального списка.

In [2]:
def remove_duplicates(arr):
    s = set()
    return (item for item in arr if item not in s and not s.add(item))

In [3]:
a=[1, 3, 2, 1, 5, 3, 5, 1, 4]

In [4]:
list(remove_duplicates(a))

[1, 3, 2, 5, 4]

Какая асимптотическая сложность у этой функций?

### O(N)

# Задача 2

Дана таблица employees всех сотрудников компании. Поля:
    full_name TEXT (PK),
    position TEXT,
    department TEXT.

Напиши запрос, выводящий все отделы, в которых меньше 5 разработчиков (position = 'Software Developer').


#### SELECT department from employees  where position = "Software Developer" GROUP BY department HAVING COUNT(position) < 5

# Задача 3
Подкинули монету N раз. Кол-во случаев, когда выпал орёл, на 10% больше, чем кол-во случаев, когда выпала решка. При каком N мы можем сказать, что монета «нечестная» (орёл и решка выпадают с разной вероятностью)?

In [5]:
def find_n(a, b, value, n=100):
    p=1
    while p > value:
        n+=1
        actual = np.array([n*a, n*b])
        expected = np.array([0.5, 0.5]) * np.sum(actual)
        _, p = chisquare(actual, expected)
    return n

In [6]:
find_n(0.45, 0.55, 0.05)

385

H0 - нет различий между ожидаемыми и действительными частотами

При N=385 мы можем отклонить H0 в пользу альтернативной, если уровень значимости α = 0.05